File: Server/Academia/models/MaterialMap.js
const mongoose = require('mongoose');
const Joi = require('joi');
const { v4: uuidv4 } = require('uuid');

/// MaterialMap Schema
const materialMapSchema = new mongoose.Schema({
    _id: {
        type: String,
        unique: true,
        default: uuidv4,
    },
    materialType: {
        type: Number,
        required: true,
    },
    materialUrl: {
        type: String,
        required: true,
        trim: true,
    },
    materialDescription: {
        type: String,
        required: true,
        trim: true,
    },
}, {
    _id: false,
    timestamps: {
        currentTime: () => new Date().toISOString(), // Use ISO 8601 format for timestamps
        createdAt: 'createdOn',
        updatedAt: 'editedOn',
    },
    collection: 'MaterialMap',
});

/// MaterialMap Model
const materialMapModel = mongoose.model('MaterialMap', materialMapSchema);

/// MaterialMap Validation
function validateCreateMaterialMap(materialMap) {
    const schema = Joi.object({
        materialType: {
            type: Joi.number().required(),
        },
        materialUrl: {
            type: Joi.string().required().trim(),
        },
        materialDescription: {
            type: Joi.string().required().trim(),
        },
    });
    return schema.validate(materialMap);
}
function validateUpdateMaterialMap(materialMap) {
    const schema = Joi.object({
        materialType: {
            type: Joi.number().required(),
        },
        materialUrl: {
            type: Joi.string().required().trim(),
        },
        materialDescription: {
            type: Joi.string().required().trim(),
        },
    });
    return schema.validate(materialMap);
}

/// MaterialMap Verbs
const createMaterialMapVerb = {
    "id-enum": "create-material-map",
    display: {
        "en-US": "Created Material Map"
    }
};
const updateMaterialMapVerb = {
    "id-enum": "update-material-map",
    display: {
        "en-US": "Updated Material Map"
    }
};
const deleteMaterialMapVerb = {
    "id-enum": "delete-material-map",
    display: {
        "en-US": "Deleted Material Map"
    }
};

module.exports = {
    materialMapModel,
    validateCreateMaterialMap,
    validateUpdateMaterialMap,
    createMaterialMapVerb,
    updateMaterialMapVerb,
    deleteMaterialMapVerb,
}

File: Mobile App/academia/lib/domain/models/material_map.dart
class MaterialMap {
  String id;
  int materialType;
  String materialUrl;
  String materialDescription;
  DateTime createdOn;
  DateTime editedOn;

  MaterialMap({
    required this.id,
    required this.materialType,
    required this.materialUrl,
    required this.materialDescription,
    required this.createdOn,
    required this.editedOn,
  });

  factory MaterialMap.fromJson(Map<String, dynamic> json) {
    return MaterialMap(
      id: json['_id'],
      materialType: json['materialType'],
      materialUrl: json['materialUrl'],
      materialDescription: json['materialDescription'],
      createdOn: DateTime.parse(json['createdOn']),
      editedOn: DateTime.parse(json['editedOn']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'materialType': materialType,
      'materialUrl': materialUrl,
      'materialDescription': materialDescription,
      'createdOn': createdOn.toIso8601String(),
      'editedOn': editedOn.toIso8601String(),
    };
  }
}